### ISSUE TITLE
(Idea) FyneDesk edition

### ISSUE THREAD

**Issue by probonopd** [furybsd/furybsd-livecd#216]
FyneDesk is a BSD licensed desktop written in Go. And it runs on FreeBSD:

![image](https://user-images.githubusercontent.com/2480569/95026225-2faf3180-0690-11eb-840c-081384bb28e3.png)

Source: https://github.com/fyne-io/fynedesk/issues/132

>> In a discussion over at #freebsd on freenode the question came up whether it would be possible to build a GNU/GPL-free desktop using FyneDesk. Would it? Or are there some hard dependencies on anything GNU/GPL?

@Fyne_io:

> Yes that should be possible :) Everything in the Fyne project is BSD licensed. In terms of GNU that is harder to answer definitively but FyneDesk is modular so we can switch in/out sections if required.

Source: https://twitter.com/Fyne_io/status/1312848007629594628?s=20

**Comment by andydotxyz** [comment]
If you'd like to check out BSD support you'll need to be on the `develop` branch which will release as 0.2 later this year.

**Comment by pkgdemon** [comment]
Very interesting.  I will definitely check this out as soon as I can.  Thanks for sharing.

**Comment by probonopd** [comment]
When trying to run

```
pkg install -y git go pkgconf glfw

cd /tmp

go get fyne.io/fynedesk
cd ~/go/src/fyne.io/fynedesk/
git checkout develop
 
go build ./cmd/fynedesk_runner || (echo "Failed to build fynedesk_runner"; exit 1)
go build ./cmd/fynedesk || (echo "Failed to build fynedesk"; exit 1)
```

then I get

```
# github.com/go-gl/glfw/v3.3/glfw
In file included from ../../../pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20200625191551-73d3c3675aa3/native_linbsd.go:10:
In file included from ../../../pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20200625191551-73d3c3675aa3/glfw/include/GLFW/glfw3native.h:114:
In file included from /usr/local/include/GL/glx.h:32:
/usr/local/include/GL/gl.h:61:11: warning: 'GLAPIENTRY' macro redefined [-Wmacro-redefined]
../../../pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20200625191551-73d3c3675aa3/glfw/include/GLFW/glfw3.h:5863:10: note: previous definition is here
install: /usr/local/furybsd/uzip/usr/local/share/xsessions/fynedesk.desktop: No such file or directory
```

Do you have any idea why this might be?

**Comment by andydotxyz** [comment]
The error is the last line, install command is failing. Perhaps that directory it is configured to write to does not exist?

**Comment by andydotxyz** [comment]
Actually the error and the build commands don’t line up - was a step from the command line missed in the pasted output/commands?

**Comment by probonopd** [comment]
No... I think the real error is the `GLAPIENTRY` thing?

**Comment by andydotxyz** [comment]
Seems unlikely it says “warning” not “error”.

**Comment by probonopd** [comment]
Indeed, those are just warnings but the binary gets built nevertheless. :+1:

**Comment by probonopd** [comment]
...and I have, at least locally, a Live ISO:

![VirtualBox_Live_06_10_2020_22_30_29](https://user-images.githubusercontent.com/2480569/95256877-b7797500-0823-11eb-9206-8edff5a15268.png)

Looks like the window manager and the bar on the right are missing though. What might be causing this?

**Comment by andydotxyz** [comment]
Looks like a min-size issue. Perhaps the desktop requires more space than it should - the current minimum is set to something that seemed sensible. How small is the screen you are using?
https://github.com/fyne-io/fynedesk/blob/2a7391451dfb2c0c37bef9516e41285335b2de28/internal/ui/desk.go#L90

**Comment by andydotxyz** [comment]
Also, how is that booted? is this through a login session, or running the embedded mode in another WM for testing?

**Comment by probonopd** [comment]
`slim` starts `/usr/local/bin/start-fyne` which in turn starts `fynedesk_launcher`.

Preliminary findings:

* Boot the ISO (I use a machine with Intel graphics and a 1080p screen)
* Ctrl-Alt-F1
* log in as liveuser (no password)
* sudo su
* service slim stop
* ps ax (fynedesk_runner is still running!)
* killall -9 fynedesk_runner
* ps ax (fynedesk_runner should no more be running!)
* startx
* In xterm, enter furybsd-xorg-tool
* Install Intel drivers (needs network connectivity)
* When it says that Xorg is restarted it will drop to a shell, enter `service slim start` there
* A FyneDesk desktop appears but it does not make use of the full 1080p resolution

**Comment by andydotxyz** [comment]
Well goodness me that was an epic day of discover.
So... the cause is that slim sets the DISPLAY to "unix:0.0" whereby "unix" is supposed to be the hostname.

I suspect that the hostname notation is either not working, or it only works when TCP connections are on for the Xserver, which they correctly are not.
However - to get this to work I edited the 'start-fynedesk' script to include 'export DISPLAY=:0.0' before executing fynedesk and it seems OK :).
I have not tested autologin yet as this was done with auto turned off so I had a better chance of seeing what was going on. And now it's late. But I think it should all work.

Even just adding the 'env' should do it:
```sh
env DISPLAY=:0.0 fynedesk_runner
```

**Comment by andydotxyz** [comment]
OK, I found a spec that mentions unix:0.0 as a special case.
I have opened a PR against the upstream X lib we use:
https://github.com/BurntSushi/xgb/pull/53

In the meantime setting DISPLAY in the script will work just fine.

**Comment by andydotxyz** [comment]
Yay, it's merged :) I will update the FyneDesk develop branch to use that version

**Comment by andydotxyz** [comment]
This is coming along very nicely :) 
Screenshot below is latest from the `feature/bsd` branch, that should be on develop in a few days.
Also installed the calculator and beebui demo apps which look pretty nice ;)

<img width="912" alt="Screenshot 2020-10-09 at 23 32 05" src="https://user-images.githubusercontent.com/294436/95636828-0de6ed80-0a88-11eb-9149-125a3132c187.png">

**Comment by probonopd** [comment]
> Also installed the calculator and beebui demo apps which look pretty nice ;)

Can you propose some code to plug in around [here](https://github.com/furybsd/furybsd-livecd/blob/11402956c14686d7955c2111d9a517b35f2af68e/settings/script.fynedesk#L44), which would give us the additional applications?

**Comment by andydotxyz** [comment]
> Known issues:

> * Resolutions under 1080p lead to non-working desktop

Resolved :)

> * Multiple of fynedesk_runner get started for unknown reasons

Resolved :)

> * Even if Xorg is killed, fynedesk keeps running for unknown reasons

Tracked as a new issue: fyne-io/fynedesk#137

**Comment by andydotxyz** [comment]
Once we merge the `feature/bsd` (PR is open at https://github.com/fyne-io/fynedesk/pull/136) You should also get:

* Fix to fyne-io/fynedesk#137
* Support for battery status
* Support for volume control

At this point I think that the FyneDesk experience on FreeBSD should be up to speed with the Linux setup.
Thanks to FreeBSD a lot of the device support was easier than I expected!
